packages/react/src/PageLayout/usePaneWidth.ts

```diff
@@ -0,0 +1,323 @@
+export const updateAriaValues = (
+  handle: HTMLElement | null,
+  values: {current?: number; min?: number; max?: number},
+) => {
+  if (!handle) return
+  if (values.min !== undefined) handle.setAttribute('aria-valuemin', String(values.min))
+  if (values.max !== undefined) handle.setAttribute('aria-valuemax', String(values.max))
+  if (values.current !== undefined) {
+    handle.setAttribute('aria-valuenow', String(values.current))
+    handle.setAttribute('aria-valuetext', `Pane width ${values.current} pixels`)
+  }
+}
+
+// ----------------------------------------------------------------------------
+// Hook
+
+/**
+ * Manages pane width state with localStorage persistence and viewport constraints.
+ * Handles initialization from storage, clamping on viewport resize, and provides
+ * functions to save and reset width.
+ */
+export function usePaneWidth({
+  width,
+  minWidth,
+  resizable,
+  widthStorageKey,
+  paneRef,
+  handleRef,
+}: UsePaneWidthOptions): UsePaneWidthResult {
+  // Derive constraints from width configuration
+  const isCustomWidth = isCustomWidthOptions(width)
+  const minPaneWidth = isCustomWidth ? parseInt(width.min, 10) : minWidth
+  const customMaxWidth = isCustomWidth ? parseInt(width.max, 10) : null
+
+  // Cache the CSS variable value to avoid getComputedStyle during drag (causes layout thrashing)
+  // Updated on mount and resize when breakpoints might change
+  const maxWidthDiffRef = React.useRef(DEFAULT_MAX_WIDTH_DIFF)
+
+  // Calculate max width constraint - for custom widths this is fixed, otherwise viewport-dependent
+  const getMaxPaneWidth = React.useCallback(() => {
+    if (customMaxWidth !== null) return customMaxWidth
+    const viewportWidth = window.innerWidth
+    return viewportWidth > 0 ? Math.max(minPaneWidth, viewportWidth - maxWidthDiffRef.current) : minPaneWidth
+  }, [customMaxWidth, minPaneWidth])
+
+  // --- State ---
+  // Current width for React renders (ARIA attributes). Updates go through saveWidth() or clamp on resize.
+  //
+  // NOTE: We read from localStorage during initial state to avoid a visible resize flicker
+  // when the stored width differs from the default. This causes a React hydration mismatch
+  // (server renders default width, client renders stored width), but we handle this with
+  // suppressHydrationWarning on the Pane element. The mismatch only affects the --pane-width
+  // CSS variable, not DOM structure or children.
+  const [currentWidth, setCurrentWidth] = React.useState(() => {
+    const defaultWidth = getDefaultPaneWidth(width)
+
+    if (!resizable || !canUseDOM) {
+      return defaultWidth
+    }
+
+    try {
+      const storedWidth = localStorage.getItem(widthStorageKey)
+      if (storedWidth !== null) {
+        const parsed = Number(storedWidth)
+        if (!isNaN(parsed) && parsed > 0) {
+          return parsed
+        }
+      }
+    } catch {
+      // localStorage unavailable - keep default
+    }
+
+    return defaultWidth
+  })
```

Comment:
Keeping this here for now, but I think we should move to a layout effect
do we have a way to preseed this into SSR to avoid hydration mismatches - or do we always want to accept a CLS flash (like we do today?)